aten/src/ATen/TensorNames.h

```diff
@@ -0,0 +1,64 @@
+#pragma once
+
+#include <ATen/core/EnableNamedTensor.h>
+#include <ATen/WrapDimUtils.h>
+
+namespace at { namespace namedinference {
+
+#ifdef BUILD_NAMEDTENSOR
+
+// TensorName and TensorNames are wrappers around Dimname and DimnameList
+// that contain helper functions to make writing name inference rules easier.
+//
+// A TensorName represents a Dimname associated with some DimnameList (from a Tensor).
+// This encapsulates all the information that is needed to check if names *match*
+// and to *unify* names.
+//
+// Definition: Two names in two tensors *match* if they are equal, or if at
+// least one of them is a wildcard that can be *refined* to the other name.
+//
+// Here is an example of checking if two names match.
+// tensor: Tensor[A, None]
+// other: Tensor[A]
+//
+// Let's say we wish to check if tensor.names[-1] matches other.names[-1].
```

Comment:
it seems a bit funny to me to use -1 for other since it only has one dim, but I guess that makes sense since we're usually going from the right?

Reply:
Yup, because we're going from the right I wanted to write this example as going from the right.